UsdAbc: translate the property's arrayExtent metadata into elementSize

cameronwhite opened the pull request:
### Description of Change(s)

For the example file in the referenced bug report, the Alembic property has the metadata `arrayExtent: 3, podExtent: 1, podName: float32_t`.
This was translated into a `float[]` array with too many elements since the `arrayExtent` was ignored. This metadata seems to have an equivalent meaning to USD's elementSize, describing a grouping of consecutive elements in the array.

Houdini's Alembic exporter, for example, uses this in the generic case for translating float tuple attributes (which might have an arbitrary tuple size) if they're not a standard type like positions or colors.

(Note that if the property instead had metadata `arrayExtent:1, podExtent:3`, this behaved correctly as-is and would translate into a `float3[]` array)

### Link to proposal ([if applicable](https://openusd.org/release/contributing_to_usd.html#step-1-get-consensus-for-major-changes))

N/A

### Fixes Issue(s)

Fixes: #3362

### Checklist

- [x] I have created this PR based on the dev branch

- [x] I have followed the [coding conventions](https://openusd.org/release/api/_page__coding__guidelines.html)

- [x] I have added unit tests that exercise this functionality (Reference: 
  [testing guidelines](https://openusd.org/release/api/_page__testing__guidelines.html))

- [x] I have verified that all unit tests pass with the proposed changes

- [x] I have submitted a signed Contributor License Agreement (Reference: 
  [Contributor License Agreement instructions](https://openusd.org/release/contributing_to_usd.html#contributor-license-agreement))

cameronwhite committed:
UsdAbc: translate arrayExtent metadata into elementSize

These values are equivalent and describe how consecutive array elements
should be grouped.

This is used, for example, by Houdini's Alembic exporter to translate
float tuples that aren't standard data types such as positions or colors.

Fixes: #3362

jesschimein commented:
Filed as internal issue #USD-11727

(This is an automated message. See [here](https://openusd.org/release/contributing_to_usd.html#github-issues) for more information.)

spiffmon commented:
@cameronwhite , while I can't be sure, it seems like `arrayExtent` is more general than just primvars in alembic?  Last year we introduced [`arraySizeConstraint` metadata for general array types in USD](https://openusd.org/release/api/class_usd_attribute.html#Usd_AttributeArraySizeConstraintAPI), and would like to promote it as preferred over `elementSize` (since it is also more expressive).

Could we ask that when reading in Alembic, you also add `arraySizeConstraint` to the attribute spec, and when writing from USD to alembic, check for that first and translate it to arrayExtent if present?

cameronwhite committed:
UsdAbc: support 'arraySizeConstraint' metadata in addition to 'elementSize'

- When reading an Alembic file, author arraySizeConstraint in addition
  to elementSize

- When writing to Alembic, prefer the arraySizeConstraint metadata over
  elementSize

cameronwhite commented:
> Could we ask that when reading in Alembic, you also add `arraySizeConstraint` to the attribute spec, and when writing from USD to alembic, check for that first and translate it to arrayExtent if present?

Thanks for the suggestion - I wasn't aware `arraySizeConstraint` existed :)
I've updated the PR with those changes
